core/src/main/java/com/linecorp/armeria/client/ClientFactory.java

```diff
@@ -201,7 +206,50 @@ static void disableShutdownHook() {
      * {@link ClientFactory} does not know how to handle the {@link ClientBuilderParams} for the provided
      * {@code client}, it should return {@link Optional#empty()}.
      */
-    <T> Optional<ClientBuilderParams> clientBuilderParams(T client);
+    default <T> Optional<ClientBuilderParams> clientBuilderParams(T client) {
+        return unwrap(client, ClientBuilderParams.class);
+    }
+
+    /**
+     * Unwraps the specified {@code client} object into the object of the specified {@code type}. For example,
+     * <pre>{@code
+     * HttpClient client = new HttpClientBuilder()
+     *     .decorator(LoggingClient.newDecorator())
+     *     .build();
+     * LoggingClient unwrapped1 = client.as(LoggingClient.class);
+     * LoggingClient unwrapped2 = ClientFactory.DEFAULT.unwrap(client, LoggingClient.class);
+     * assert unwrapped1 == unwrapped2;
+     * }</pre>
+     *
+     * @param client the client object
+     * @param type the type of the object to return
+     * @return the object of the specified {@code type} if found. {@link Optional#empty()} if not found.
+     */
+    default <T> Optional<T> unwrap(Object client, Class<T> type) {
```

Comment:
Don't we need a test for this?

Reply:
`clientBuilderParams()` cover this method, so I didn't write one, but yeah I can add one.

Reply:
Added tests, found and fixed some bugs :scream: